File: week13/15/2_CarPool.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        int passengerNum[1001] = {0, };
        for (int i = 0; i < trips.size(); i++){
            passengerNum[trips[i][1]] += trips[i][0];
            passengerNum[trips[i][2]] -= trips[i][0];
        }
        int CurPassenger = 0;
        for(int i = 0; i < 1001; i++) {
            CurPassenger += passengerNum[i];
            if (CurPassenger > capacity) return false;
        }
        return true;

    }
};

File: week13/15/2_CarPool.py
import heapq
from typing import List

class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        trips.sort(key = lambda t: t[1])

        minHeap = []
        curpassenger = 0

        for trip in trips:
            passengerNum, start, end = trip
            while minHeap and minHeap[0][0] <= start:
                curpassenger -= minHeap[0][1] 
                heapq.heappop(minHeap)
            curpassenger += passengerNum
            if curpassenger > capacity:
                return False
            heapq.heappush(minHeap, [end, passengerNum])
        return True
        